cpp/stats_server.cc

```diff
@@ -0,0 +1,266 @@
+                                 metadata_hostname->second.length())
+                  << std::endl;
+      }
+    } else {
+      std::cout << status.error_code() << ": " << status.error_message()
+                << std::endl;
+    }
+    user_ = response.name();
+    // User requested premium service, but the user is not a premium user,
+    // reject the request.
+    if (membership_ == "premium" &&
+        response.membership() != MembershipType::PREMIUM) {
+      std::cout << "token: " << token_ << ", name: " << user_
+                << ", membership: " << response.membership() << ","
+                << std::endl;
+      std::cout << "requested membership: " << membership_
+                << ", authentication FAILED" << std::endl;
+      return false;
+    }
+    std::cout << "token: " << token_ << ", name: " << user_
+              << ", membership: " << response.membership() << "," << std::endl;
+    std::cout << "requested membership: " << membership_
+              << ", authentication success true" << std::endl;
+    return true;
+  }
+
+  Status FetchPrice(ServerContext* context, const PriceRequest* request,
+                    PriceResponse* response) override {
+    if (!ObtainAndValidateUserAndMembership(context)) {
+      return Status(StatusCode::UNAUTHENTICATED,
+                    "membership authentication failed");
+    }
+    context->AddInitialMetadata("hostname", hostname_);
+    response->set_price(sin(time(0) * 1000 / 173) * 1000 + 10000);
+    return Status::OK;
+  }
+
+  Status WatchPrice(ServerContext* context, const PriceRequest* request,
+                    ServerWriter<PriceResponse>* writer) override {
+    if (!ObtainAndValidateUserAndMembership(context)) {
+      return Status(StatusCode::UNAUTHENTICATED,
+                    "membership authtication failed");
+    }
+    context->AddInitialMetadata("hostname", hostname_);
+    while (true) {
+      PriceResponse response;
+      response.set_price(sin(time(0) * 1000 / 173) * 1000 + 10000);
+      if (!writer->Write(response)) {
+        break;
+      }
+      std::this_thread::sleep_for(
+          std::chrono::milliseconds(membership_ == "premium" ? 100 : 1000));
+    }
+    return Status::OK;
+  }
+
+  std::string hostname_;
+  bool premium_only_ = false;
+  std::unique_ptr<Account::Stub> account_stub_;
+  std::string token_;
+  std::string user_ = "Alice";
+  std::string membership_ = "premium";
+};
+
+void RunServer(const std::string& port, const std::string& account_server,
+               const std::string& hostname_suffix, const bool premium_only) {
+  char base_hostname[256];
+  if (gethostname(base_hostname, 256) != 0) {
+    std::cout << "unable to get host name" << std::endl;
+    return;
+  }
+  std::string hostname(base_hostname);
+  hostname += hostname_suffix;
+  std::string server_address("0.0.0.0:");
+  server_address += port;
+  StatsServiceImpl service;
+  service.SetHostName(hostname);
+  service.SetPremiumOnly(premium_only);
```

Comment:
Make a constructor for `(hostname, premium, stub)` instead of `set`?

Reply:
done